rclcpp/include/rclcpp/strategies/message_pool_memory_strategy.hpp

```diff
@@ -68,16 +79,24 @@ class MessagePoolMemoryStrategy
    */
   std::shared_ptr<MessageT> borrow_message()
   {
-    size_t current_index = next_array_index_;
-    next_array_index_ = (next_array_index_ + 1) % Size;
-    if (pool_[current_index].used) {
-      throw std::runtime_error("Tried to access message that was still in use! Abort.");
+    std::lock_guard<std::mutex> lock(pool_mutex_);
+    if (free_list_.size() == 0) {
+      throw std::runtime_error("No more free slots in the pool");
     }
-    pool_[current_index].msg_ptr_->~MessageT();
-    new (pool_[current_index].msg_ptr_.get())MessageT;
 
-    pool_[current_index].used = true;
-    return pool_[current_index].msg_ptr_;
+    size_t current_index = free_list_.pop_front();
+
+    return std::shared_ptr<MessageT>(
+      pool_[current_index], [this](MessageT * p) {
+        for (size_t i = 0; i < Size; ++i) {
+          if (pool_[i] == p) {
+            p->~MessageT();
+            new(p) MessageT();
```

Comment:
It feels weird to call the constructor here. Imo, it should either:

- call the constructor on borrow, or
- never call the constructor after the first time

Reply:
Both of these are problematic, I think.

> * call the constructor on borrow, or

We could do this, but then we'd have to be more careful in the destructor of `MessagePoolMemoryStrategy`.  In particular, you can't call `delete` on an object that has already had the destructor run (at least, valgrind gives me warnings when I do this).  So we'd have to do additional tracking to not try to `delete` these (and even that doesn't work, since then we'd leak the memory).

> * never call the constructor after the first time

This I don't think we can do because it will break semantics of the messages.  In particular, assume you had a `std::vector` in one of the messages (which is what we generate for unbounded arrays).  So during the first borrow, the user code pushes 2 items onto the `std::vector`, then returns it to the pool.  Now the next borrow that uses that slot gets an object that already has 2 items in the vector.  Yes, this would be a weird usage, but I think we should always return freshly-initialized objects to the user code.

The current code basically always ensures that we have a freshly-initialized object ready to hand out in the pool.  And to me, that seems to be the easiest to reason about.  Thoughts?

Reply:
Despite the need to be careful with the destructor of the pool, I think the first option I suggested is better. Consider also that the constructor of the messages might throw (`std::bad_alloc` for example), and you'd be doing that from the destructor of the shared pointer, rather than from the borrow call.

For the second option, I agree we'd probably want to return clean messages, but for example in the case of using a borrowed message with take, it always resets all the fields anyway, so it might be a small performance improvement to avoid the destructor/constructor in some cases (users do this implicitly with a message on the stack in a class with a publish loop sometimes). Maybe the answer is to destruct then construct in the borrow function with an optional flag to skip that step (with understanding that the user will be responsible for dealing with stale data in the message).

Either way, I really think at least construction should be done in the borrow call.

Reply:
> Either way, I really think at least construction should be done in the borrow call.

All right, that's fair.  In 7901ef3, I ended up revamping this so we only do allocation/deallocation in the `MessagePoolMemoryStrategy` constructor and destructor, respectively.  We then do placement-new construction when we hand out the `shared_ptr` from the pool, and manually run the destructor when the custom `shared_ptr` deleter is called.  I think this should satisfy all of the constraints here; please take another look.